plenum/server/replica.py

```diff
@@ -1688,6 +1706,87 @@ def can_pp_seq_no_be_in_view(self, view_no, pp_seq_no):
 
+    def request_pre_prepare_if_possible(self, three_pc_key) -> bool:
+        """
+        Check if has an acceptable PRE_PREPARE already stashed, if not then
+        check count of PREPAREs, make sure >f consistent PREPAREs are found,
+        store the acceptable PREPARE state (digest, roots) for verification of
+        the received PRE-PREPARE
+        """
+        if len(self.preparesWaitingForPrePrepare[three_pc_key]) < self.quorums.prepare.value:
+            logger.debug('{} not requesting a PRE-PREPARE because does not have'
+                         ' sufficient PREPAREs for {}'.format(self, three_pc_key))
+            return False
+
+        if three_pc_key in self.requested_pre_prepares:
+            logger.debug('{} not requesting a PRE-PREPARE since already '
+                         'requested for {}'.format(self, three_pc_key))
+            return False
+
+        if three_pc_key in self.prePreparesPendingPrevPP:
+            logger.debug('{} not requesting a PRE-PREPARE since already found '
+                         'stashed for {}'.format(self, three_pc_key))
+            return False
+
+        # Choose a better data structure for `prePreparesPendingFinReqs`
+        pre_prepares = [pp for pp, _, _ in self.prePreparesPendingFinReqs
+                        if (pp.viewNo, pp.ppSeqNo) == three_pc_key]
+        digest, state_root, txn_root, prepare_senders = \
+            self.get_acceptable_stashed_prepare_state(three_pc_key)
+        if pre_prepares:
+            if [pp for pp in pre_prepares if
+                (pp.digest, pp.stateRootHash, pp.txnRootHash) == (digest, state_root, txn_root)]:
+                logger.debug('{} not requesting a PRE-PREPARE since already '
+                             'found stashed for {}'.format(self, three_pc_key))
+                return False
+
+        # TODO: Using a timer to retry would be a better thing to do
+        logger.debug('{} requesting PRE-PREPARE({}) from {}'.
+                     format(self, three_pc_key, prepare_senders))
+        self.node.request_msg(PREPREPARE, {f.INST_ID.nm: self.instId,
+                                           f.VIEW_NO.nm: three_pc_key[0],
+                                           f.PP_SEQ_NO.nm: three_pc_key[1]},
+                              [self.getNodeName(s) for s in prepare_senders])
+        self.requested_pre_prepares[three_pc_key] = digest, state_root, txn_root
+        return True
+
+    def get_acceptable_stashed_prepare_state(self, three_pc_key):
+        prepares = {s: (m.digest, m.stateRootHash, m.txnRootHash) for m, s in
+                    self.preparesWaitingForPrePrepare[three_pc_key]}
+        acceptable = mostCommonElement(prepares.values())
+        return (*acceptable, {s for s, state in prepares.items()
+                              if state == acceptable})
+
+    def process_requested_pre_prepare(self, pp: PrePrepare, sender: str):
```

Comment:
The PRE-PREPARE is requested from all nodes sending PREPARE (2f) so even if one nodes gives the correct PRE-PREPARE, it's done